### ISSUE TITLE
CDK2 example failing on osx

### ISSUE THREAD

**Issue by jchodera** [choderalab/perses#471]
With `openeye-toolkits==2018.2.1` and `ambertools==17.6`, I'm getting the following failure when running the CDK2 example:
```
lski2414:cdk2-example choderaj$ python ../../scripts/setup_relative_calculation.py cdk2_setup.yaml
Warning: OEAssignPartialCharges has been deprecated please use OEAssignCharges instead!
Warning: SelectElfDiverseConfs: elfPop.NumConfs 1 <= elfLimit 1
DEBUG:openmoltools.openeye:keep_confs was set to 1. Molecule positions will be reset.
DEBUG:openmoltools.openeye:~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine
/private/var/folders/m1/wl3ptzqx2csf2r_q2bfy78n0h7kcfv/T/tmpvxvstsia/molecule0000000000.tripos.mol2
Traceback (most recent call last):
  File "../../scripts/setup_relative_calculation.py", line 35, in <module>
    setup_dict = relative_setup.run_setup(setup_options)
  File "/Users/choderaj/miniconda/lib/python3.6/site-packages/perses-0.1-py3.6.egg/perses/dispersed/relative_setup.py", line 1009, in run_setup
    solvate=solvate, atom_map=atom_map)
  File "/Users/choderaj/miniconda/lib/python3.6/site-packages/perses-0.1-py3.6.egg/perses/dispersed/relative_setup.py", line 111, in __init__
    ffxml = forcefield_generators.generateForceFieldFromMolecules(mol_list)
  File "/Users/choderaj/miniconda/lib/python3.6/site-packages/openmoltools/forcefield_generators.py", line 437, in generateForceFieldFromMolecules
    run_antechamber(prefix, input_mol2_filename, charge_method=None, net_charge=net_charge, gaff_mol2_filename=gaff_mol2_filename, frcmod_filename=frcmod_filename, gaff_version=gaff_version)
  File "/Users/choderaj/miniconda/lib/python3.6/site-packages/openmoltools/amber.py", line 373, in run_antechamber
    raise Exception(msg)
Exception: antechamber failed to produce output mol2 file
command: antechamber -i in.mol2 -fi mol2 -o out.mol2 -fo mol2 -s 2 -at gaff -nc 0
output:
--------------------------------------------------------------------------------

Welcome to antechamber 17.3: molecular input file processor.

acdoctor mode is on: check and diagnosis problems in the input file.
-- Check Format for mol2 File --
   Status: pass
Info: Finished reading file (in.mol2).
-- Check Unusual Elements --
   Status: pass
-- Check Open Valences --
   Status: pass
-- Check Geometry --
      for those bonded   
      for those not bonded   
   Status: pass
-- Check Weird Bonds --
/Users/choderaj/miniconda/bin/to_be_dispatched/antechamber: Fatal Error!
Weird atomic valence (2) for atom (ID: 18, Name: N2).
       Possible open valence.--------------------------------------------------------------------------------
input mol2:
--------------------------------------------------------------------------------
@<TRIPOS>MOLECULE
~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine
   45    48     0     0     0
SMALL
USER_CHARGES

@<TRIPOS>ATOM
      1 C1          5.5575   -4.6963   -6.3837 C.ar      1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.9290
      2 C2          6.1041   -6.5796   -2.3816 C.2       1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.2789
      3 C3          2.5481   -2.6819   -4.6160 C.3       1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.1433
      4 C4          1.4655   -2.0383   -3.7516 C.3       1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine        -0.0645
      5 C5          2.0913   -1.2929   -2.5650 C.3       1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine        -0.0758
      6 C6          1.0315   -0.5958   -1.7150 C.3       1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine        -0.0742
      7 C7          0.1597    0.3340   -2.5532 C.3       1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine        -0.0771
      8 C8         -0.4652   -0.3984   -3.7366 C.3       1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine        -0.0742
      9 C9          0.5920   -1.0953   -4.5900 C.3       1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine        -0.0758
     10 C10         4.3838   -2.9203  -10.8595 C.ar      1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine        -0.1294
     11 C11         3.8025   -1.7515  -10.3681 C.ar      1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine        -0.0986
     12 C12         3.8693   -1.4587   -9.0059 C.ar      1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine        -0.0256
     13 C13         4.5173   -2.3349   -8.1352 C.ar      1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine        -0.0970
     14 N1          4.5407   -4.0206   -5.7666 N.ar      1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine        -0.7812
     15 C14         4.3045   -4.2763   -4.4468 C.ar      1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.7396
     16 O1          3.2846   -3.6025   -3.8252 O.3       1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine        -0.4491
     17 C15         5.1087   -5.2078   -3.8040 C.ar      1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine        -0.2786
     18 N2          5.0087   -5.6151   -2.3911 N.2       1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine        -0.3203
     19 N3          6.7300   -6.6958   -3.6169 N.pl3     1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine        -0.4014
     20 C16         6.1249   -5.8430   -4.5350 C.ar      1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.4675
     21 N4          6.4061   -5.6232   -5.8645 N.ar      1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine        -0.7468
     22 N5          5.7525   -4.3870   -7.7491 N.pl3     1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine        -0.7713
     23 C17         5.0985   -3.5038   -8.6267 C.ar      1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.1481
     24 C18         5.0317   -3.7965   -9.9888 C.ar      1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine        -0.1103
     25 H1          6.4722   -7.1871   -1.5730 H         1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.0744
     26 H2          3.2335   -1.9214   -5.0018 H         1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.0457
     27 H3          2.0992   -3.2194   -5.4566 H         1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.0457
     28 H4          0.8273   -2.8406   -3.3577 H         1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.0491
     29 H5          2.6547   -1.9940   -1.9376 H         1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.0522
     30 H6          2.8087   -0.5464   -2.9294 H         1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.0522
     31 H7          0.3998   -1.3497   -1.2292 H         1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.0366
     32 H8          1.5170   -0.0238   -0.9161 H         1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.0366
     33 H9         -0.6301    0.7634   -1.9265 H         1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.0396
     34 H10         0.7674    1.1697   -2.9215 H         1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.0396
     35 H11        -1.1842   -1.1407   -3.3685 H         1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.0366
     36 H12        -1.0277    0.3119   -4.3533 H         1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.0366
     37 H13         1.2201   -0.3370   -5.0751 H         1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.0522
     38 H14         0.0972   -1.6568   -5.3915 H         1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.0522
     39 H15         4.3317   -3.1482  -11.9201 H         1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.1407
     40 H16         3.3012   -1.0794  -11.0600 H         1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.1498
     41 H17         4.5728   -2.1112   -7.0718 H         1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.1579
     42 H18         7.5120   -7.3115   -3.8214 H         1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.3185
     43 H19         6.5128   -4.9057   -8.1758 H         1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.4669
     44 H20         5.4810   -4.7042  -10.3825 H         1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine         0.1457
     45 Br1         3.0808    0.1251   -8.3410 Br        1 ~{N}-(3-bromophenyl)-6-(cyclohexylmethoxy)purin-9-ium-7-id-2-amine        -0.0839
@<TRIPOS>BOND
     1    1   14 ar
     2    1   21 ar
     3    1   22 1
     4    2   18 1
     5    2   19 2
     6    2   25 1
     7    3    4 1
     8    3   16 1
     9    3   26 1
    10    3   27 1
    11    4    5 1
    12    4    9 1
    13    4   28 1
    14    5    6 1
    15    5   29 1
    16    5   30 1
    17    6    7 1
    18    6   31 1
    19    6   32 1
    20    7    8 1
    21    7   33 1
    22    7   34 1
    23    8    9 1
    24    8   35 1
    25    8   36 1
    26    9   37 1
    27    9   38 1
    28   10   11 ar
    29   10   24 ar
    30   10   39 1
    31   11   12 ar
    32   11   40 1
    33   12   13 ar
    34   12   45 1
    35   13   23 ar
    36   13   41 1
    37   14   15 ar
    38   15   16 1
    39   15   17 ar
    40   17   18 1
    41   17   20 ar
    42   19   20 1
    43   19   42 1
    44   20   21 ar
    45   22   23 1
    46   22   43 1
    47   23   24 ar
    48   24   44 1
--------------------------------------------------------------------------------
```
When I write out the molecule read by `relative_setup.py:load_sdf()`, it looks like the bicyclic aromatic scaffold is no longer being perceived as aromatic after being read in. I'm not quite sure why this is, but I'm investigating!

**Comment by pgrinaway** [comment]
Hm, thanks for investigating! I recall once that we had an issue with having `ambertools` instead of `ambermini` installed. I don't remember what exactly the problem was, though.

**Comment by ivyzhang999** [comment]
@jchodera did you ever figure this out? I'm having a similar error with one of the kinase inhibitors: Pazopanib,CC1=C(C=C(C=C1)NC2=NC=CC(=N2)N(C)C3=CC4=NN(C(=C4C=C3)C)C)S(=O)(=O)N

One of the bonds that should be aromatic is not being perceived as such (this is in `gaffTemplateGenerator`'s `generateResidueTemplate`), so I'm getting a weird open valence of 5 for that carbon atom.

However, when I write a dummy script reading in the above smiles, converting it to an oemol, and then writing it to a mol2 file, the aromaticity is correct.

**Comment by jchodera** [comment]
Where precisely are we relying on `generateResidueTemplate()` to generate parameters for us? I thought @pgrinaway had switched to always generating `ffxml` parameters using the `OEMol` directly, as in here:

https://github.com/choderalab/perses/blob/master/perses/dispersed/relative_setup.py#L111

**Comment by ivyzhang999** [comment]
In the KinaseInhibitor testsystem, when we add the solvent using openMM's `addSolvent` method, another openMM method `createSystem` is called, which relies on the Forcefield object's template generators. This is why openmoltools's `gaffTemplateGenerator` is called.

Here's the traceback: 
Traceback (most recent call last):
  File "test_ivy.py", line 13, in <module>
    testsystem = KinaseInhibitorsTestSystem(storage_filename="kinase_ivy.nc")
  File "/home/zhangi/perses/perses/tests/testsystems.py", line 1827, in __init__
    super(KinaseInhibitorsTestSystem, self).__init__(**kwargs)
  File "/home/zhangi/perses/perses/tests/testsystems.py", line 1733, in __init__
    modeller.addSolvent(forcefield, model='tip3p', padding=9.0*unit.angstrom)
  File "/home/zhangi/miniconda3/envs/cenv/lib/python3.6/site-packages/simtk/openmm/app/modeller.py", line 376, in addSolvent
    system = forcefield.createSystem(self.topology)
  File "/home/zhangi/miniconda3/envs/cenv/lib/python3.6/site-packages/simtk/openmm/app/forcefield.py", line 1129, in createSystem
    if generator(self, res):
  File "/home/zhangi/openmoltools/openmoltools/forcefield_generators.py", line 582, in gaffTemplateGenerator
    [template, ffxml] = generateResidueTemplate(molecule)
  File "/home/zhangi/openmoltools/openmoltools/forcefield_generators.py", line 366, in generateResidueTemplate
    run_antechamber(template_name, input_mol2_filename, charge_method=None, net_charge=net_charge, gaff_mol2_filename=gaff_mol2_filename, frcmod_filename=frcmod_filename, gaff_version=gaff_version)
  File "/home/zhangi/openmoltools/openmoltools/amber.py", line 373, in run_antechamber
    raise Exception(msg)
Exception: antechamber failed to produce output mol2 file

**Comment by jchodera** [comment]
The `ForceField.createSystem` method will use whatever residue templates and parameters we have already registered in the `ForceField` object. If we modify the `KinaseInhibitorsTestSystem` to first process the `OEMol` objects corresponding to the inhibitors you want to use to produce `ffxml` fiels that can be added to `ForceField`, the templated will already be available when `createSystem()` is called.

Can you point me to the code where `KinaseInhibitorsTestSystem` appears?

**Comment by ivyzhang999** [comment]
I talked to Patrick and updated KinaseInhibitorTestSystem as you suggested, however the issue with Pazopanib persists. I'm going to ignore it for now, but if you have any other ideas let me know!

**Comment by jchodera** [comment]
> I talked to Patrick and updated KinaseInhibitorTestSystem as you suggested, however the issue with Pazopanib persists. I'm going to ignore it for now, but if you have any other ideas let me know!

Can you point me to the code? It's hard for me to help debug without being able to read through the code and try it out!

**Comment by ivyzhang999** [comment]
Yes, sorry about that! 
Here is KinaseInhibitorTestSystem: https://github.com/choderalab/perses/blob/master/perses/tests/testsystems.py#L1799
Here is the line in the base class that adds solvent: https://github.com/choderalab/perses/blob/master/perses/tests/testsystems.py#L1721

**Comment by jchodera** [comment]
To fix this and pre-generated parameters, we would want to
* no longer register `gaffTemplateGenerator` in [this line](https://github.com/choderalab/perses/blob/master/perses/tests/testsystems.py#L1710)
* pregenerate the `ffxml` from the OEMols by adding something based on [this line](https://github.com/choderalab/perses/blob/master/perses/dispersed/relative_setup.py#L111)
* load that `ffxml` into the `forcefield` with:
```python
forcefield.loadFile(StringIO(ffxml))
```

With these changes, `ForceField` will no longer try to generate parameters on the fly, but will instead pregenerate the parameters for all molecules (which will take a little while). If we wanted to speed this up, we could pre-generate the `ffxml` file and cache it to disk. We can get more complicated if we like.

**Comment by jchodera** [comment]
This will be fixed by #521
